gneiss/sort.py

```diff
@@ -0,0 +1,109 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016--, gneiss development team.
+#
+# Distributed under the terms of the GPLv3 License.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+# ----------------------------------------------------------------------------
+import numpy as np
+import pandas as pd
+from functools import partial
+from gneiss.util import match
+
+
+def mean_niche_estimator(abundances, gradient):
+    """ Estimates the mean niche of an organism.
+
+    Calculates the mean niche of an organism along a gradient.
+    This is done by calculating the expected value of an organism
+    across the gradient.
+
+    Specifically, this module calculates the following
+
+    .. math::
+        E[g | x] =
+        \frac{1}{\sum\limits_{j=1}^N x_j} \sum\limits_{i=1}^N x_i g_i
+
+    Where :math:`N` is the number of samples, :math:`x_i` is the proportion of
+    species :math:`x` in sample :math:`i`, :math:`g_i` is the gradient value
+    at sample `i`.
+
+    Parameters
+    ----------
+    abundances : pd.Series, np.float
+        Vector of fraction abundances of an organism over
+        a list of samples.
+    gradient : pd.Series, np.float
+        Vector of numerical gradient values.
+
+    Returns
+    -------
+    np.float :
+        The mean gradient that the organism lives in.
+
+    Raises
+    ------
+    ValueError:
+        If the length of `abundances` is not the same length as `gradient`.
+    ValueError:
+        If the length of `gradient` contains nans.
+    """
+    len_abundances = len(abundances)
+    len_gradient = len(gradient)
+    if len_abundances != len_gradient:
+        raise ValueError("Length of `abundances` (%d) doesn't match the length"
+                         " of the `gradient` (%d)" % (len_abundances,
+                                                      len_gradient))
+    if np.any(pd.isnull(gradient)):
+        raise ValueError("`gradient` cannot have any nans.")
+
+    # normalizes the proportions of the organism across all of the
+    # samples to add to 1.
+    v = abundances / abundances.sum()
+    m = np.dot(gradient, v)
```

Comment:
Yup

On Jul 27, 2016 11:56 AM, "Antonio Gonzalez" <notifications@github.com>
wrote:

> In gneiss/sort.py
> <https://github.com/biocore/gneiss/pull/16#discussion_r72501098>:
>
> > +    ValueError:
> > +        If the length of `gradient` contains nans.
> > +    """
> > +    len_abundances = len(abundances)
> > +    len_gradient = len(gradient)
> > +    if len_abundances != len_gradient:
> > +        raise ValueError("Length of `abundances` (%d) doesn't match the length"
> > +                         " of the `gradient` (%d)" % (len_abundances,
> > +                                                      len_gradient))
> > +    if np.any(pd.isnull(gradient)):
> > +        raise ValueError("`gradient` cannot have any nans.")
> > +
> > +    # normalizes the proportions of the organism across all of the
> > +    # samples to add to 1.
> > +    v = abundances / abundances.sum()
> > +    m = np.dot(gradient, v)
>
> Is this line really doing this:
> E[g | x] = \frac{1}{\sum\limits_{j=1}^N x_j} \sum\limits_{i=1}^N x_i g_i
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/biocore/gneiss/pull/16/files/23acedfd2291c0b3716023c9867c2629b1c16dc4#r72501098>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AD_a3cfBc1UiQuuaW3PR3kUGYTYwh3Fcks5qZ6m_gaJpZM4JUnpZ>
> .
>